### PR TITLE
Connect frontend to eFolder

### PR DESCRIPTION
**Pull request by mdbenjam**
Connects: #2340

To integrate with eFolder Express the Caseflow client needs to request documents from the eFolder Express backend. This PR introduces a feature hidden behind the `efolder_docs_api` feature flag that pulls document content from eFolder Express.

**Test Plan**
- Run Caseflow on port 3000, and eFolder on port 4000. Don't enable the feature flag and make sure we get the documents from Caseflow.
- Run Caseflow on port 3000, and eFolder on port 4000. Enable the feature flag and make sure we get the documents from eFolder.
    - Enable the feature flag by executing `FeatureToggle.enable!(:efolder_docs_api)` in the rails console for Caseflow

### CODE REVIEW

File: client/app/reader/reducer.js [hunk 1]
```diff
@@ -157,6 +157,7 @@ const initialShowErrorMessageState = {
 
 export const initialState = {
   assignments: [],
+  efolderDocumentUrl: () => {},
```
Comment (NickHeiner) [inline client/app/reader/reducer.js]: In Redux, it's best for the state to be fully serializable. This allows us to do things like time-travel debugging, debug dumps from users with their state history, Jest snapshot reducer tests, etc. 

A JS function is not serializable, so this breaks that constraint.

There is no need to put a function in the state. Just store `action.payload.url`, and have a util function that uses that URL to produce the final endpoint.

If you want to make it easy to share this function across multiple components, you can use a `selector` (found in `selectors.js`) to make a function that takes in the Redux state and returns the `efolderDocumentUrl` function. That way, from the component's perspective, it essentially looks the same – it still shows up as a function on `props`.

Comment (NickHeiner) [inline client/app/reader/reducer.js]: Also, functions should be named with verbs in them. (I recognize that documentUrl, which this is replacing, probably came from me. My bad.) This would be better named as `getEfolderDocumentUrl`.

File: client/app/components/Pdf.jsx [hunk 2]
```diff
@@ -463,23 +463,24 @@ export class Pdf extends React.PureComponent {
       return Promise.resolve(this.predrawnPdfs[file].pdfDocument);
     }
 
-    return PDFJS.getDocument(file).then((pdfDocument) => {
-      if ([...this.props.prefetchFiles, this.props.file].includes(file)) {
+    return PDFJS.getDocument({ url: file,
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: This would typically be formatted as:

```js
return PDFJS.getDocument({ 
      url: file,
      withCredentials: true 
    }).then((pdfDocument) => {
```

File: client/app/components/Pdf.jsx [hunk 3]
```diff
@@ -463,23 +463,26 @@ export class Pdf extends React.PureComponent {
       return Promise.resolve(this.predrawnPdfs[file].pdfDocument);
     }
 
-    return PDFJS.getDocument(file).then((pdfDocument) => {
-      if ([...this.props.prefetchFiles, this.props.file].includes(file)) {
+    return PDFJS.getDocument({ 
+      url: file,
+      withCredentials: true 
+    }).then((pdfDocument) => {
+        if ([...this.props.prefetchFiles, this.props.file].includes(file)) {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: The indentation on this and the following lines is off. `npm run lint:fix` will reveal the problems.

File: client/app/reader/PdfViewer.jsx [hunk 4]
```diff
@@ -130,13 +129,16 @@ export class PdfViewer extends React.Component {
 
   selectedDocId = () => Number(this.props.match.params.docId)
 
+  prevDoc = () => _.get(this.props.documents, [this.selectedDocIndex() - 1])
```
Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: I know you're following the pattern with `prevDocId`, but it's better to name functions with a verb. Let's name this `getPrevDoc`. (Same comment applies on the next line.)

Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: And as long as we're going to have these functions, it would be nice to refactor `prevDocId` and `nextDocId` to use them. (You don't have to rename `prevDocId` and `nextDocId` now.)

File: client/app/reader/PdfViewer.jsx [hunk 5]
```diff
@@ -130,13 +129,16 @@ export class PdfViewer extends React.Component {
...
   prevDocId = () => _.get(this.props.documents, [this.selectedDocIndex() - 1, 'id'])
   nextDocId = () => _.get(this.props.documents, [this.selectedDocIndex() + 1, 'id'])
 
   getPrefetchFiles = () => _.compact([
-    this.prevDocId(),
-    this.nextDocId()
-  ]).map(documentPath)
+    this.prevDoc(),
+    this.nextDoc()
+  ]).map((doc) => doc.content_url)
```
Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: This can just be:

```
.map('content_url')
```

File: client/app/reader/ReaderLoadingScreen.jsx [hunk 6]
```diff
@@ -43,7 +41,7 @@ export class ReaderLoadingScreen extends React.Component {
         };
 
         for (let i = 0; i < PARALLEL_DOCUMENT_REQUESTS; i++) {
-          downloadDocuments(documents.map((doc) => documentUrl(doc)), i);
+          downloadDocuments(documents.map((doc) => doc.content_url), i);
```
Comment (NickHeiner) [inline client/app/reader/ReaderLoadingScreen.jsx]: This can be:

```js
downloadDocuments(_.map(documents, 'content_url'), i)
```

**Comment by NickHeiner** [comment]
Is it too much of a pain to add a test for this? :smile:

File: client/app/reader/PdfViewer.jsx [hunk 7]
```diff
@@ -130,13 +129,13 @@ export class PdfViewer extends React.Component {
...
+  getNextDoc = () => _.get(this.props.documents, [this.selectedDocIndex() + 1])
 
-  getPrefetchFiles = () => _.compact([
-    this.prevDocId(),
-    this.nextDocId()
-  ]).map(documentPath)
+  getPrevDocId = () => _.get(this.getPrevDoc(), 'id')
+  getNextDocId = () => _.get(this.getNextDoc(), 'id')
+
+  getPrefetchFiles = () => _.map([this.getPrevDoc(), this.getNextDoc()], 'content_url')
```
Comment (mogrenAtWork) [inline client/app/reader/PdfViewer.jsx]: Lodash is magical

File: app/models/document.rb [hunk 8]
```diff
@@ -155,6 +156,14 @@ def serialized_receipt_date
     serialize_date(receipt_date)
   end
 
+  def content_url
+    if FeatureToggle.enabled?(:efolder_docs_api) && RequestStore.store[:application] == "reader"
+      URI(ExternalApi::EfolderService.efolder_base_url + "/api/v1/documents/#{efolder_id}").to_s
```
Comment (mogrenAtWork) [inline app/models/document.rb]: I changed this since URI.join will remove any paths appended to `ExternalApi::EfolderService.efolder_base_url` (i.e. if `ExternalApi::EfolderService.efolder_base_url` is http://vets.gov/caseflow, calling `URI.join('https://vets.gov/caseflow', '/api/v1') will produce 'https://vets.gov/api/v1'

Using `URI()` will ensure `ExternalApi::EfolderService.efolder_base_url` is a valid URL

File: config/initializers/efolder.rb [hunk 9]
```diff
@@ -1,3 +1,10 @@
 # Fakes::VBMSService provides the same functionality as a Fakes::EfolderService would. We can refactor
 # this later when we switch Caseflow to read all documents from efolder permanently
-EFolderService = (!ApplicationController.dependencies_faked? ? ExternalApi::EfolderService : Fakes::VBMSService)
+
+# If you're in a development or test environment and want to use the EfolderService, just set the environment variable
+# EFOLDER_EXPRESS_URL
```
Comment (mogrenAtWork) [inline config/initializers/efolder.rb]: This env variable is read in config/application.rb.

ExternalApi::EfolderService.efolder_base_url reads this configuration. I call the EfolderService method so we don't have to change where we're retrieving this value in a bunch of different places (if it ever changes)

File: lib/fakes/appeal_repository.rb [hunk 10]
```diff
@@ -314,20 +314,14 @@ def self.seed_certification_data!
...
-
     [
       Generators::Document.build(vbms_document_id: 1, type: "NOD", category_procedural: true),
       Generators::Document.build(vbms_document_id: 2, type: "SOC", category_medical: true),
       Generators::Document.build(vbms_document_id: 3, type: "Form 9",
                                  category_medical: true, category_procedural: true),
       Generators::Document.build(
         vbms_document_id: 5,
-        type: super_long_title,
+        type: "This is a very long document type let's see what it does to the UI!",
```
Comment (mogrenAtWork) [inline lib/fakes/appeal_repository.rb]: This was just a way for Artem to test the previous changes that the flags to enable the Efolder API were working

File: config/initializers/efolder.rb [hunk 11]
```diff
@@ -1,3 +1,10 @@
...
+EFolderService = if !ApplicationController.dependencies_faked? || !ExternalApi::EfolderService.efolder_base_url.empty?
```
Comment (NickHeiner) [inline config/initializers/efolder.rb]: Since we have an `if / else`, can we reverse it and not need the negative condition? I think that would make it easier to read.

File: config/initializers/efolder.rb [hunk 12]
```diff
@@ -1,3 +1,10 @@
 # Fakes::VBMSService provides the same functionality as a Fakes::EfolderService would. We can refactor
 # this later when we switch Caseflow to read all documents from efolder permanently
-EFolderService = (!ApplicationController.dependencies_faked? ? ExternalApi::EfolderService : Fakes::VBMSService)
+
+# If you're in a development or test environment and want to use the EfolderService, just set the environment variable
```
Comment (NickHeiner) [inline config/initializers/efolder.rb]: Can you put this instruction in the README?

**Comment by NickHeiner** [comment]
I have done some smoke testing for the tests failing when the code is broken.

File: client/app/reader/ReaderLoadingScreen.jsx [hunk 13]
```diff
@@ -37,13 +35,13 @@ export class ReaderLoadingScreen extends React.Component {
             return;
           }
 
-          ApiUtil.get(documentUrls[index], { cache: true }).then(
+          ApiUtil.get(documentUrls[index], { cache: true, withCredentials: true }).then(
```
Comment (mogrenAtWork) [inline client/app/reader/ReaderLoadingScreen.jsx]: This will pass the correct cookie information when making AJAX calls and prefetching documents.

**Comment by mogrenAtWork** [comment]
I'm still seeing `Refused to get unsafe header "Accept-Ranges"` from pdf-worker (i.e. Pdf.js) and I'm not sure how they're being generated

**Comment by mogrenAtWork** [comment]
@NickHeiner this is ready for ya

**Comment by NickHeiner** [comment]
Code LGTM. I'm going to do a final check that it works locally once you fix the merge conflicts, then we'll be ready to go.

File: spec/models/document_spec.rb [hunk 14]
```diff
@@ -37,6 +38,59 @@
     end
   end
 
+  context ".content_url" do
+    context "and efolder_docs_api is enabled and application is reader" do
+      before do
+        FeatureToggle.enable!(:efolder_docs_api)
```
Comment (shanear) [inline spec/models/document_spec.rb]: may want to disable this after the test, it could affect other tests down the road

File: spec/models/document_spec.rb [hunk 15]
```diff
@@ -37,6 +38,59 @@
...
+        RequestStore.store[:application] = "reader"
+        expect(ExternalApi::EfolderService).to receive(:efolder_base_url).and_return(base_url).once
```
Comment (shanear) [inline spec/models/document_spec.rb]: `EfolderService` vs `EFolderService`?

File: spec/models/document_spec.rb [hunk 16]
```diff
@@ -37,6 +38,59 @@
...
+      end
+
+      let(:base_url) { Faker::Internet.url }
+
+      it "returns the URL for the document in efolder" do
+        document.efolder_id = Random.rand(999_999_999)
```
Comment (shanear) [inline spec/models/document_spec.rb]: we have a method that already does this that you might be able to use here `Generators::Base.generate_external_id`

File: spec/models/document_spec.rb [hunk 17]
```diff
@@ -37,6 +38,59 @@
...
+        expect(document.content_url).to eq(base_url + "/api/v1/documents/#{document.efolder_id}")
+      end
+    end
+
+    context "and efolder_docs_api is enabled and application is not reader" do
```
Comment (shanear) [inline spec/models/document_spec.rb]: any reason for clumping the context conditions rather than doing something like:

```
context "efolder_docs_api is enabled" do
  context "application is reader" do
    #..
  end

  context "application is not reader" do
    #..
  end
end
```

File: spec/models/document_spec.rb (hunk 15 above)
Comment (mogrenAtWork) [inline spec/models/document_spec.rb]: I tried naming it `EFolderService` but rails could not find the file, `efolder_service.rb` as a result

File: spec/models/document_spec.rb (hunk 17 above)
Comment (mogrenAtWork) [inline spec/models/document_spec.rb]: Fixed

File: spec/models/document_spec.rb (hunk 16 above)
Comment (mogrenAtWork) [inline spec/models/document_spec.rb]: 👍

File: spec/models/document_spec.rb (hunk 14 above)
Comment (mogrenAtWork) [inline spec/models/document_spec.rb]: Done

File: client/app/components/PdfUI.jsx [hunk 18]
```diff
@@ -279,7 +279,6 @@ PdfUI.propTypes = {
     type: PropTypes.string,
     receivedAt: PropTypes.string
```
Comment (shanear) [inline client/app/components/PdfUI.jsx]: needs `content_url`?

File: client/app/reader/PdfViewer.jsx [hunk 19]
```diff
@@ -130,13 +129,13 @@ export class PdfViewer extends React.Component {
...
+  getNextDoc = () => _.get(this.props.documents, [this.selectedDocIndex() + 1])
 
-  getPrefetchFiles = () => _.compact([
-    this.prevDocId(),
-    this.nextDocId()
-  ]).map(documentPath)
+  getPrevDocId = () => _.get(this.getPrevDoc(), 'id')
+  getNextDocId = () => _.get(this.getNextDoc(), 'id')
+
+  getPrefetchFiles = () => _.compact(_.map([this.getPrevDoc(), this.getNextDoc()], 'content_url'))
```
Comment (shanear) [inline client/app/reader/PdfViewer.jsx]: Is this refactoring related to the `content_url` change?
